tree_arena/src/tree_arena_unsafe.rs

```diff
@@ -0,0 +1,609 @@
+    /// Item in the node
+    pub item: &'arena mut T,
+    /// Children of the node
+    pub children: ArenaMutChildren<'arena, T>,
+}
+
+/// A handle giving mutable access to an arena item's children.
+///
+/// See [`ArenaMut`] for more information.
+///
+/// This stores all the permissions for what nodes can be accessed from the current node
+/// As such if a [`std::mem::swap`] is used to swap the children of two trees,
+/// each tree will still have the correct permissions. This also stores the roots, and so
+/// that will also be in a consistent state
+#[derive(Debug)]
+pub struct ArenaMutChildren<'arena, T> {
+    /// The associated data arena
+    parent_arena: &'arena mut DataMap<T>,
+    /// The parent id for these children
+    id: Option<NodeId>,
+    /// Array of children
+    child_arr: &'arena mut Vec<NodeId>,
+}
+
+impl<Item> Clone for ArenaRef<'_, Item> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<Item> Copy for ArenaRef<'_, Item> {}
+
+impl<T> Clone for ArenaRefChildren<'_, T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<Item> Copy for ArenaRefChildren<'_, Item> {}
+
+impl<T> DataMap<T> {
+    fn new() -> Self {
+        Self {
+            items: HashMap::new(),
+            parents: HashMap::new(),
+        }
+    }
+
+    /// Find an item in the tree.
+    ///
+    /// Returns a shared reference to the item if present.
+    ///
+    /// Time Complexity O(1)
+    fn find(&self, id: impl Into<NodeId>) -> Option<ArenaRef<'_, T>> {
+        // SAFETY
+        // We need there to be no mutable access to the node
+        // Mutable access to the node would imply there is some &mut self
+        // As we are taking &self, there can be no mutable access to the node
+        // Thus this is safe
+        let id: NodeId = id.into();
+        let parent_id = *self.parents.get(&id)?;
+        self.items
+            .get(&id)
+            .and_then(|item| unsafe { item.get().as_ref() })
+            .map(
+                |TreeNode {
+                     item,
+                     children: child_arr,
+                 }| {
+                    let children = ArenaRefChildren {
+                        parent_arena: self,
+                        id: Some(id),
+                        child_arr,
+                    };
+                    ArenaRef {
+                        parent_id,
+                        item,
+                        children,
+                    }
+                },
+            )
+    }
+
+    /// Find an item in the tree.
+    ///
+    /// Returns a mutable reference to the item if present.
+    ///
+    /// Time Complexity O(1)
+    ///
+    /// # SAFETY
+    ///
+    /// When using this on [`ArenaMutChildren`] associated with some node,
+    /// must ensure that `id` is a descendant of that node, otherwise can
+    /// obtain two mutable references to the same node
+    unsafe fn find_mut(&mut self, id: impl Into<NodeId>) -> Option<ArenaMut<'_, T>> {
```

Comment:
Hi! I have a minor note on the safety documentation. This is sort of a comment out of nowhere, but I've developed an interest in reviewing unsafe code and I've been following the development in this repo.

Since this retains the full `&mut self` inside the returned `ArenaMutChildren`, I think the safety requirements are more extensive than the case described here. Any use of `parent_arena` in `ArenaMutChildren` needs to be careful to not invalidate the reference produced from `unsafe { item.get().as_mut() }` below. E.g. `DataMap::find` also must only be called on descendants, `parent_arena.items.remove()` must only operate on descendants, `parent_arena.items.insert()` must not overwrite existing nodes, `parent_arena.items` must be used to clear the whole structure, etc.

I was initially thinking the safety documentation could be reworked to note the overall requirement that the `parent_arena` field must not be used to invalidate the returned `item` reference and then include a few examples. However, I think all call sites would have the same safety note pointing to the details `ArenaMutChildren`. So since all the details of `ArenaMutChildren` are private to this module, `find_mut` could be safe, and the `unsafe` block in it could be documented with the fact that all operations that `ArenaMutChildren` allows only access descendants and never invalidate the item reference created here.

Reply:
Agree with the overall point that the safety requirement should be better put as that you can only access children or remove children of the current node (I'm not sure whether insert() comment would be needed as it is a panic to insert a node with the same name as the key, and not sure clearing is different to removal as there isn't a clear api, and I assume any would be an action on the Arena itself thus checked by the borrow checker as having a reference to an item in the tree prevents any mutable access of the tree itself)

Reply:
Regarding safety, I think this makes sense, but can also see another argument that it being unsafe is correct in the case it did become public (though assume then the same should be true of the immutable method otherwise could obtain a immutable and mutable reference to the same node)

Reply:
> I'm not sure whether insert() comment would be needed as it is a panic to insert a node with the same name as the key,

> there isn't a clear api, and I assume any would be an action on the Arena itself thus checked by the borrow checker as having a reference to an item in the tree prevents any mutable access of the tree itself

These are comments about what the safe code in `ArenaMutChildren` is allowed to do with the internal `items` hashmap (which has things such as insert without such checks and clear).